shared/types/indexer_test.go

```diff
@@ -0,0 +1,120 @@
+	// check indexing/get by hash
+	hash, err := txResult.Hash()
+	require.NoError(t, err)
+	txResultFromHash, err := txIndexer.GetByHash(hash)
+	require.NoError(t, err)
+	require.True(t, txResultsEqual(t, txResult, txResultFromHash))
+	// check indexing/get by height
+	txResultsFromHeight, err := txIndexer.GetByHeight(0)
+	require.NoError(t, err)
+	txResultsFromHeight1, err := txIndexer.GetByHeight(1)
+	require.NoError(t, err)
+	require.Equal(t, 2, len(txResultsFromHeight))
+	require.Equal(t, 1, len(txResultsFromHeight1))
+	require.True(t, txResultsEqual(t, txResult, txResultsFromHeight[0]))
+	require.True(t, txResultsEqual(t, txResult2, txResultsFromHeight[1]))
+	require.True(t, txResultsEqual(t, txResult3, txResultsFromHeight1[0]))
+	// check indexing by sender / recipient
+	sender := txResult3.GetSender()
+	recipient := txResult3.GetRecipient()
+	require.NoError(t, err)
+	txResultsFromSender, err := txIndexer.GetBySender(sender, false)
+	require.NoError(t, err)
+	txResultsFromRecipient, err := txIndexer.GetByRecipient(recipient, false)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(txResultsFromSender))
+	require.Equal(t, 1, len(txResultsFromRecipient))
+	require.True(t, txResultsEqual(t, txResult3, txResultsFromSender[0]))
+	require.True(t, txResultsEqual(t, txResult3, txResultsFromRecipient[0]))
+	// ensure it's not indexed elsewhere
+	txResultsFromSenderBad, err := txIndexer.GetBySender(recipient, false)
+	require.NoError(t, err)
+	txResultsFromRecipientBad, err := txIndexer.GetByRecipient(sender, false)
+	require.NoError(t, err)
+	require.Equal(t, 0, len(txResultsFromSenderBad))
+	require.Equal(t, 0, len(txResultsFromRecipientBad))
+}
+
+func txResultsEqual(t *testing.T, txR1, txR2 TxResult) bool {
+	bz, err := txR1.Bytes()
+	require.NoError(t, err)
+	bz2, err := txR2.Bytes()
+	require.NoError(t, err)
+	return bytes.Equal(bz, bz2)
+}
+
+// utility helpers
+
+func NewTestingTransactionResult(height, index int) TxResult {
+	testingTransaction := randLetterBytes()
+	resultCode, err := randomErr()
+	return &DefaultTxResult{
+		Tx:          testingTransaction,
+		Height:      int64(height),
+		Index:       int32(index),
+		ResultCode:  resultCode,
+		Error:       err,
+		Signer:      randomAddress(),
+		Recipient:   randomAddress(),
+		MessageType: randomMessageType(),
+	}
+}
+
+func randomMessageType() string {
+	msgTypes := []string{"send", "stake", "unstake", "editStake", "unjail"}
+	return msgTypes[rand.Intn(len(msgTypes))]
+}
+
+func randomAddress() string {
+	add, _ := crypto.GenerateAddress()
+	return add.String()
+}
+
+func randomErr() (code int32, err string) {
+	errors := []string{"insufficient funds", "address not valid", "invalid signature"}
+	code = int32(0)
+	err = ""
+	if rand.Intn(4) == 1 {
+		code = int32(rand.Intn(len(errors)))
+		err = errors[code]
+	}
+	return
+}
+
+func randLetterBytes() []byte {
+	rand.Seed(int64(time.Now().Nanosecond()))
+	letters := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
```

Comment:
for my understanding, why are we using `rune` instead of `byte` here?

Reply:
Generally speaking, `rune` is regarded as the `char` primitive of a string. In this case, I think byte would also be acceptable since it's ascii chars

Reply:
Noted. Found this online: https://www.callicoder.com/golang-basic-types-operators-type-conversion

**tl;dr `rune` is UTF8 and `char` is ASCII only.**

Feel free to resolve.